### PR TITLE
Check plugin configuration when changes occur

### LINKED ISSUE #45: Check the generated webhook secret
Sometimes, the webhook secret contains a `/` character. This make the router to be confused, and the webhooks never arrive.
We should check the webhook secret value before validation
image

### LINKED ISSUE #81: Prevent plugin from starting when At Rest Encryption Key is not set
It is currently possible to run the plugin when `At Rest Encryption Key` is not set. 

I'm not sure we should allow this. If a user attempts to connect to their GitHub account in this case they will see `Internal error when storing your token`
Logs show
```
{"level":"dpanic","ts":1616790113.21193,"caller":"zap@v1.16.0/sugar.go:191","msg":"Ignored key without a value.","plugin_id":"com.github.mattermost.plugin-circleci","ignored":"crypto/aes: invalid key size 0\ncould not create a cipher block, check key\ngithub.com/mattermost/mattermost-plugin-circleci/server/store.encrypt\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/store/utils.go:34\ngithub.com/mattermost/mattermost-plugin-circleci/server/store.(*Store).StoreTokenForUser\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/store/kv_store.go:37\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).executeAccountConnect\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/account.go:154\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).executeAccount\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/account.go:56\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).ExecuteCommand\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/command.go:222\ngithub.com/mattermost/mattermost-server/v5/plugin.(*hooksRPCServer).ExecuteCommand\n\t/Users/dylanhaussermann/go/pkg/mod/github.com/mattermost/mattermost-server/v5@v5.27.0/plugin/client_rpc_generated.go:114\nreflect.Value.call\n\t/usr/local/go/src/reflect/value.go:476\nreflect.Value.Call\n\t/usr/local/go/src/reflect/value.go:337\nnet/rpc.(*service).call\n\t/usr/local/go/src/net/rpc/server.go:377\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1371\nError occurred while encrypting access token\ngithub.com/mattermost/mattermost-plugin-circleci/server/store.(*Store).StoreTokenForUser\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/store/kv_store.go:39\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).executeAccountConnect\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/account.go:154\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).executeAccount\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/account.go:56\ngithub.com/mattermost/mattermost-plugin-circleci/server/plugin.(*Plugin).ExecuteCommand\n\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-circleci/server/plugin/command.go:222\ngithub.com/mattermost/mattermost-server/v5/plugin.(*hooksRPCServer).ExecuteCommand\n\t/Users/dylanhaussermann/go/pkg/mod/github.com/mattermost/mattermost-server/v5@v5.27.0/plugin/client_rpc_generated.go:114\nreflect.Value.call\n\t/usr/local/go/src/reflect/value.go:476\nreflect.Value.Call\n\t/usr/local/go/src/reflect/value.go:337\nnet/rpc.(*service).call\n\t/usr/local/go/src/net/rpc/server.go:377\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1371"}
```
```
{"level":"error","ts":1616790113.2118974,"caller":"mlog/sugar.go:25","msg":"Error when storing token","plugin_id":"com.github.mattermost.plugin-circleci"}
```
0/5 we may want to just prevent the startup of the plugin in this case

### PR DESCRIPTION
**Pull request by nathanaelhoun**
## Changes

- [x] Check plugin configuration when it changes and at startup
    - [x] At Rest Encryption Key is defined (closes #81)
    - [x] Webhook Token is defined and does not contains a forwardslash (closes #45).
- [x] Plugin fails to start if a configuration error is detected
- [x] Added unit tests
- [x] Fix missing keys for values in `API.LogError()` calls (mentionned in #81) 

## Changelog

```
## Enhancements
- Plugin configuration is checked at startup, and plugin fails to start if an error is detected
```

### CODE REVIEW

File: server/plugin/configuration.go [hunk 1]
```diff
@@ -86,6 +84,14 @@ func (p *Plugin) OnConfigurationChange() error {
 		return errors.New("please provide the Webhook Secret")
 	}
 
+	if strings.Contains(configuration.WebhooksSecret, "/") {
+		return errors.New("please regenerate the Webhook Secret, it mustn't contains a '/'")
```
Comment (jfrerich) [inline server/plugin/configuration.go]:
```suggestion
		return errors.New("please regenerate the Webhook Secret, it cannot contain a '/'")
```

File: server/plugin/manifest.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ const manifestStr = `
         "key": "WebhooksSecret",
         "display_name": "Webhooks Secret:",
         "type": "generated",
-        "help_text": "The secret used to authenticate the webhook to Mattermost.",
+        "help_text": "The secret used to authenticate the webhook to Mattermost. Should not contain a forwardslash.",
```
Comment (jfrerich) [inline server/plugin/manifest.go]:
```suggestion
        "help_text": "The secret used to authenticate the webhook to Mattermost. Should not contain a forward slash.",
```

File: plugin.json [hunk 3]
```diff
@@ -24,7 +24,7 @@
                 "key": "WebhooksSecret",
                 "display_name": "Webhooks Secret:",
                 "type": "generated",
-                "help_text": "The secret used to authenticate the webhook to Mattermost.",
+                "help_text": "The secret used to authenticate the webhook to Mattermost. Should not contain a forwardslash.",
```
Comment (jfrerich) [inline plugin.json]:
```suggestion
                "help_text": "The secret used to authenticate the webhook to Mattermost. Should not contain a forward slash.",
```

**Review by jfrerich — CHANGES_REQUESTED** [review]
Thanks, @nathanaelhoun! I've got just a handfull of nitpicks!

**Comment by nathanaelhoun** [comment]
Thanks for the review @jfrerich, suggestions applied :wink:

**Review by jfrerich — APPROVED** [review]
Thanks @nathanaelhoun!

**Comment by nathanaelhoun** [comment]
Hi! Any news on this?

**Comment by hanzei** [comment]
@nathanaelhoun Heads up that there is a merge conflict

**Comment by nathanaelhoun** [comment]
Thanks for the heads-up @hanzei, there are now fixed!

**Comment by jfrerich** [comment]
@nathanaelhoun!  We have been in a major crunch to release the apps framework and that pushed back some of the bounty plugin work. @DHaussermann do are you ok with mering this PR and testing post-merge, similar to what you suggested in https://github.com/mattermost/mattermost-plugin-circleci/pull/79?

**Review by DHaussermann — APPROVED** [review]
I agree with @jfrerich 

This is suitable to test as part of release testing post-merge

LGTM!

**Comment by nathanaelhoun** [comment]
Thanks for the review @jfrerich @DHaussermann. 
The apps framework looks really exciting :smile:

**Pull request merged**